de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorManipulator.java

```diff
@@ -233,4 +218,77 @@ public void adjustViewport(Editor editor, LineRange range, TextSelection selecti
 
     // todo: implement actual viewport adjustment logic
   }
+
+  /**
+   * Replaces the content of the given file with the given content. This is done on the Document
+   * level.
+   *
+   * <p><b>NOTE:</b> This method should only be used as part of the recovery process.
+   *
+   * @param path the path of the file to recover
+   * @param content the new content of the file
+   * @param encoding the encoding of the content
+   * @param source the user that send the recovery action
+   * @see Document
+   * @see SharedResourcesManager#handleFileRecovery(FileActivity)
+   */
+  public void handleContentRecovery(SPath path, byte[] content, String encoding, User source) {
+    VirtualFile virtualFile = VirtualFileConverter.convertToVirtualFile(path);
+    if (virtualFile == null) {
+      return;
+    }
+
+    Document document = projectAPI.getDocument(virtualFile);
+    if (document == null) {
+      return;
+    }
+
+    int documentLength = document.getTextLength();
+
+    IFile file = path.getFile();
+    annotationManager.removeAnnotations(file);
+
+    String text;
+    try {
+      text = new String(content, encoding);
+    } catch (UnsupportedEncodingException e) {
+      LOG.warn("Could not decode text using given encoding. Using default instead.", e);
+
+      text = new String(content);
```

Comment:
The chances are good that this trigger another inconsistency.

Reply:
Yes, but I don't know what else we should/could do in this situation. As the encoding is specified by the client (who *should* have created the byte array with the given encoding), this case *should* not be possible.

Do you have a suggestion what else to do in this case?

Reply:
Done.